Refactor tape recall

belforte opened the pull request:
this PR does 2 main things:
1. when a task in status TAPERECALL is killed, it goes in KILLRECALL
2. introduces a new recurring action TapeRecallManager which will replace current TapeRecallStatus and handle both TAPERECALL and KILLRECALL states

improvements:
- when it kills tasks in TAPERECALL it also removes the Rucio recall rule
- when a rule completes and task is released for submission, it extends rule lifetime for another 30 days
- less code, get rid of useless instantiation of a new MasterWorker in TapeRecallStatus
- paves the way to further code cleanup and unification with other TW places

This is not fully tested yet, but ready for a first review.

**IMPORTANT NOTE**
when we will want to test and deploy, we must change also REST or tasks will never go into KILLRECALL

belforte commented:
I tested this on `test2` and it was all OK . But just one dataset.

mapellidario commented:
>  get rid of useless instantiation of a new MasterWorker wet TapeRecallStatus

wow, thank you! :)

belforte commented:
More tests. Everything find in latest version.

belforte commented:
please test this

belforte commented:
Ready for @novicecpp to review. Code works, but I will  still be happy to get guidance on how better to structure.

novicecpp commented on src/python/TaskWorker/Actions/Recurring/TapeRecallManager.py:
```diff
@@ -0,0 +1,239 @@
+                self.logger.debug("tm_DDM_reqid' is not defined for task %s, skipping it", taskName)
+                continue
+            self.logger.info("Task points to Rucio RuleId:  %s ", reqId)
+            try:
+                rule = self.rucioClient.get_replication_rule(reqId)
+            except RuleNotFound:
+                msg = "Rucio rule id %s not found. Please report to experts" % reqId
+                self.logger.error(msg)
+                self.uploadWarning(msg, taskName)
+                continue
```
One case that can get weird, but not bug, is when `MAX_DAYS_FOR_TAPERECALL` in condition line 79 is *more than* rule expiration time due to misconfiguration. Worst case is TaskWorker keep upload same error every 4 hour.
I am not sure if we need to take any action or add a bit of comment here or it can silently ignore.

belforte replied:
I agree that it is a misconfiguration, then it is OK to raise an error so that someone reports it and we can fix it. We do not have a machinery in place to have such problems automatically reported to CRAB DevOps !  CRAB Server has a limit to how many messages it accepts so this will not flood DB nor user's console.
That's a general/generic line of thought. On this I think we can code so that it can't happen. Let me look at it.

novicecpp commented on src/python/TaskWorker/Actions/Recurring/TapeRecallManager.py:
```diff
@@ -0,0 +1,239 @@
+            taskName = aTask['tm_taskname']
+            self.logger.info("Working on task %s", taskName)
+            # 1.) check for "waited too long"
+            if (time.time() - getTimeFromTaskname(str(taskName))) > MAX_DAYS_FOR_TAPERECALL*24*60*60:
```
Not directly relate to this line but you should update https://github.com/dmwm/CRABServer/blob/55f2ac9b2458dd8cb010812b431710e9918b1fc2/src/python/ServerUtilities.py#L83 to current policy (30 days) and also use it in
https://github.com/dmwm/CRABServer/blob/55f2ac9b2458dd8cb010812b431710e9918b1fc2/src/python/TaskWorker/Actions/DBSDataDiscovery.py#L193

belforte replied:
No. It is correct as it is. But clearly not explained well enough to be understandable.
Do you think that stating current policien in the comments [here](https://github.com/dmwm/CRABServer/blob/55f2ac9b2458dd8cb010812b431710e9918b1fc2/src/python/TaskWorker/Actions/DBSDataDiscovery.py#L93-L102) can do ? Alas policies have two components: the actions (in the code like `requestTapeRecall()` ) and the value of the constants (in `ServerUtilities.py`).

novicecpp replied:
No no. I am misunderstanding this. I remember this https://github.com/dmwm/CRABServer/pull/7591 but not https://github.com/dmwm/CRABServer/pull/7612 and that why I am confused.

I saw latest commit and it looks much better to understand.

novicecpp reviewed (changes requested):
Additional questions and comments:
- Do you have a doc explaining how and why states are changed (or a  state machine diagram)? I want to understand why we need KILLFAILED but not KILLRECALLFAILED.
- `getRESTParams()` code itself is fine. But the end goal should be to make the config object ready to fetch its attribute. Checking or manipulating configuration should be done before passing it to the class.

Thanks for the changes. It looks very promising to have cleaner code in our codebase :)

belforte commented:
Thanks @novicecpp , great review.
You have two questions
>     Do you have a doc explaining how and why states are changed (or a state machine diagram)? I want to understand why we need KILLFAILED but not KILLRECALLFAILED.

 We have [this](https://twiki.cern.ch/twiki/bin/view/CMSPublic/CMSCrabTaskStates) which is alas obsolete already since it does not have TAPERECALL. It was done with a too-much-human-intensive tool, we shoudl redo in a simpler form which has a chance of being ketp current. Anyhow the thing is that KILLFAILED means "could not talk to condor scheduler to kill dagman there", but for a task in TAPERECALL there is no action, only change status from TAPERECALL to KILLED. If that were to fail, we could not set to KILLFAILED either. 
 As a second order correction, we could use KILLRECALLFAILED to indicate that we failed to remove the Rucio rule, I had not thought about that. But am not sure that we care to know and investigate those (the only use of KILLFAILED is that user may tell  us so we check what went wrong inside TW) instead of simply forget and let rule expire.
 So all in all I'd keep things as they are, with the action item to create a new stage diagram picture. 
 A better diagram is not easy. There is an original sin here. The people who initially designed things did not think in terms of FSM. So the task status field does nor specify the status, e.g. for a task in QUEUED it only means that TW is working on it, but it could be doing submission, resubmission, kill, or progress after dry run. :sob: 

>    getRESTParams() code itself is fine. But the end goal should be to make the config object ready to fetch its attribute. Checking or manipulating configuration should be done before passing it to the class.

I surely agree. But lack the imagination to figure out what exactly could be done to make this happen. Something like "first time we read config file we should create `restHost` and `dbInstance` so that they are accessed au-pair with all other params ?"

Your in-line comments are very good and true, give me some time to address them all. Thanks

belforte commented:
I tried one online tool and created the picture below. Doing that I realized that the FSM diagram in the twiki is fully obsoleted. Since I was pissed off with that "it does not only matter which status you are in, but also how you got there", I had them switch to use "status + command" as "actual status". I.e. a task can be in "NEW with Command = SUBMIT" or in "NEW with Command = KILL"

So this picture is likely not correct either (it should be carefully compared with the code), and lacks info on what caused the state transitions. It i is very difficult to get rid of original sins.
![TaskFSM](https://user-images.githubusercontent.com/1837785/235725768-89c908a8-362a-4b28-871d-5e630432bf97.png)

belforte commented on src/python/CRABInterface/DataWorkflow.py:
```diff
@@ -390,11 +390,18 @@ def kill(self, workflow, killwarning=''):
+            self.api.modify(self.Task.SetStatusWarningTask_sql, status=["NEW"], command=["KILL"],
+                            taskname=[workflow], warnings = [str(warnings)])
+
+        if row.task_status == 'TAPERECALL':
```
:scream:

belforte replied:
OMG !!!! What did I do ??? OK. OK. Thanks

belforte commented on src/python/TaskWorker/Actions/Recurring/TapeRecallManager.py:
```diff
@@ -0,0 +1,239 @@
+                expiration = rule['expires_at']  # this is a datetime.datetime object
+                # this condition could be met only if rule expiration is shorter then MAX_DAYS_FOR_TAPERECALL
+                # which is more a mistake than a possibility, anyhow, let's cover all grounds
+                if expiration < datetime.datetime.now():  # did we wait too much ?
```
Yeah.. I tool that from old code in `TapeRecallStatus`. I guess I initially thought that rule info would stick in Rucio even after rule expires. But this is not the case. Let's clean it up.

belforte commented:
try again a new FSM diagram (yeah.. this does not belong here)
![TaskFSM](https://user-images.githubusercontent.com/1837785/235731184-8ec6c8c5-9ea8-4852-8dc3-0da3a16a6a96.png)

belforte commented:
I need to make a list of action items here:
- [ ] fix https://github.com/dmwm/CRABServer/pull/7620#discussion_r1182258009 (correct use of `elif` )
- [ ] fix https://github.com/dmwm/CRABServer/pull/7620#discussion_r1182259853 (introduce RULE_LIFETIME in ServerUtilities and use also in DBSDataDiscovery)
- [ ] fix https://github.com/dmwm/CRABServer/pull/7620#discussion_r1182261164  (remove useless `if` )
- [ ] make sure that MAX_DAYS_FOR_RECALL is never larger than RULE_LIFETIME https://github.com/dmwm/CRABServer/pull/7620#discussion_r1182292925 .
   * idea: in DBSDataDiscover/requestTapeRecall set the initiall lifetime for the rule at MAX_DAYS_FOR_RECALL + 7days (7 days is some comfortable margin in case TW or TapeRecallManager is not working for a few days so that we do not loose track of pending requests). Keep RULE_LIFETIME as the value which is set when recall is complete and we set task to NEW. Maye call it 'INPUT_DATA_LOCKING_LIFETIME' ? and reuse in future general "input locking" ? Or use directly TASK_LIFETIME and do w/o a RULE_LIFETIME ?
- [ ]  explain in comments why it is ok that MAX_DAYS_FOR_RECALL is different (shorter) than RULE_LIFETIME https://github.com/dmwm/CRABServer/pull/7620#discussion_r1182295772
- [ ] map server instance to restHost+dbInstance at time of config. file read and deprecate getRESTParams()
- [ ] review Task FSM diagram, document it, and setup a zoom meeting to review this

novicecpp commented:
Thanks for answering the questions. It is clear to me now.

novicecpp commented:
> [ ] map server instance to restHost+dbInstance at time of config. file read and deprecate getRESTParams()

We should open a separate issue for this. The refactoring stuff needs some time to think about how to do it properly. My rough idea is the same as you said in https://github.com/dmwm/CRABServer/pull/7620#issuecomment-1531726093 .

> [ ] review Task FSM diagram, document it, fix TASKDBSTATUSES* in ServerUtilities.py and setup a zoom meeting to review this

My suggestion is to create diagrams as code via PlantUML https://plantuml.com/state-diagram .
It is a nice tool to generate diagram images for you. We can commit PlantUML code in git, and easy to edit later.
There is https://github.com/mikitex70/plantuml-markdown which supports generating images on the fly in `mkdocs`, but I never try it. Initially, we can render locally and commit both image and code. 
Diagram images from https://github.com/dmwm/CRABServer/issues/7463 are generated by PlantUML in my local machine (not the kube one, I draw it in draw.io).
I would be happy to help work on this.

belforte commented:
@novicecpp I think I fixed the issues which you raised in comments withing the code and moved last two boxes to separate issues.
Let me know if you have further comments or requests.

novicecpp reviewed (approved):
Thanks. LGTM.
